### ISSUE TITLE
Confusion between layers.Activation('softmax'...) and activations.softmax(...)

### ISSUE THREAD

**Issue by yourtheron** [tensorflow/tensorflow#35537]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 2.0.0
- Python version: 3.7.5
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0.1
- GPU model and memory: QuadroK620M

I am building a custom model, and trying to use mixed precision. Please take a look at the following two custom models
```
from __future__ import absolute_import, division, print_function, unicode_literals
import functools

import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, activations


class SFM1(tf.keras.Model):
    def __init__(self):
        super(SFM1, self).__init__()
        self.output_layer = layers.Activation('softmax', dtype='float32')
    
    def call(self, inputs):
        return self.output_layer(inputs)

class SFM2(tf.keras.Model):
    def __init__(self):
        super(SFM2, self).__init__()
        
    
    def call(self, inputs):
        return activations.softmax(inputs, axis=1)
    
x = tf.random.uniform((2, 4, 64, 64, 64), dtype=tf.float32)
sfm1 = SFM1()
y1 = sfm1(x)

sfm2 = SFM2()
y2= sfm2(x)

tf.math.equal(y1, y2)
```
y1 and y2 differ, because layers.Activation(...) assumes axis=-1 by default.
The issue is, that in mixed presion doc [here](https://www.tensorflow.org/guide/keras/mixed_precision#training_the_model_with_a_custom_training_loop)
The dtype of the output layer shall be specified as tf.float32, but activations.softmax(...) cannot do it. If I use layers.Activation(...), the axis cannot be specified as 1 ( I am using channel_first data, and I am afraid that cannot change source data). So what can I do?
Shall I have to work around by 
``` tf.dtypes.cast(activations.softmax(inputs, axis=1), dtype=tf.float32)```

**Comment by gowthamkpr** [comment]
@yourtheron  You can refer to the following [question](https://stackoverflow.com/questions/53503985/what-is-the-difference-between-keras-activations-softmax-and-keras-layers-softma) which should be able to answer your question.

Also please post questions like these on stackoverflow as this question is not related to bug/performance, build/install, feature request or docs related issue. Thanks!

**Comment by yourtheron** [comment]
@gowthamkpr . Thanks for the notice. And I will do so in future. But unfortunately, that post on Stackoverflow cannot answer my question. Softmax layer cannot specify precision, which means that if I use mixed precision and a Softmax layer as the output layer, I cannot convert float16 back into float32. Any work-around? I have asked the same thing in the post on Statckoverflow.

**Comment by reedwm** [comment]
First note, you only need to do softmax in float32 at the end of the model. Unless the softmax is immediately followed by the loss, it is typically safe to do in float16. This means the softmax in `SFM1` and `SFM2` is fine to do in float16 unless you plan on having those layers be the last layers of the model. Therefore you do not have to pass `dtype='float32'` to softmax.

Now to answer the question: What if you did want to pass `axis` to softmax and do it in float32? This is unlikely as most losses expect the softmax to be done on the final axis, but it can occur with custom losses. Casting the inputs to float32 with `tf.dtypes.cast` and using the `activations.softmax` function is a good solution. But your example is slightly incorrect: You want to cast the *inputs* to softmax, not the *outputs*:

```
activations.softmax(tf.dtypes.cast(inputs, dtype=tf.float32), axis=1)
```

Alternatively, you can directly use a `tf.keras.layers.Softmax` layer, which does have an activation argument:
```
layer = tf.keras.layers.Softmax(axis=1, dtype=tf.float32)
```
